src/main/scala/fi/oph/koski/valpas/repository/ValpasDatabaseService.scala

```diff
@@ -295,21 +300,29 @@ WITH
   ORDER BY
     oppivelvollinen_oppija.sukunimi,
     oppivelvollinen_oppija.etunimet
-    """)).as[(ValpasOppijaResult)])
+    """)).as[Either[HttpStatus, ValpasOppijaResult]])
   }
 
-  implicit private val getValpasOppijaResult: GetResult[ValpasOppijaResult] = GetResult(r => {
+  implicit private def getValpasOppijaResult: GetResult[Either[HttpStatus, ValpasOppijaResult]] = GetResult(r => {
     val rs: ResultSet = r.rs
-    ValpasOppijaResult(
-      henkilö = ValpasHenkilö(
-        oid = rs.getString("oppija_oid"),
-        hetu = Option(rs.getString("hetu")),
-        syntymäaika = Option(rs.getObject("syntymaaika", classOf[LocalDate])),
-        etunimet = rs.getString("etunimet"),
-        sukunimi = rs.getString("sukunimi")
-      ),
-      oikeutetutOppilaitokset = rs.getArray("oikeutetutOppilaitokset").getArray.asInstanceOf[Array[String]].toSet,
-      opiskeluoikeudet = JsonSerializer.parse[List[ValpasOpiskeluoikeus]](rs.getString("opiskeluoikeudet"))
-    )
+    parseOpiskeluoikeudet(rs.getString("opiskeluoikeudet"))
+      .map(opiskeluoikeudet =>
+        ValpasOppijaResult(
+          henkilö = ValpasHenkilö(
+            oid = rs.getString("oppija_oid"),
+            hetu = Option(rs.getString("hetu")),
+            syntymäaika = Option(rs.getObject("syntymaaika", classOf[LocalDate])),
+            etunimet = rs.getString("etunimet"),
+            sukunimi = rs.getString("sukunimi")
+          ),
+          oikeutetutOppilaitokset = rs.getArray("oikeutetutOppilaitokset").getArray.asInstanceOf[Array[String]].toSet,
+          opiskeluoikeudet = opiskeluoikeudet
+        )
+      )
   })
+
+  private def parseOpiskeluoikeudet(jsonResult: String): Either[HttpStatus, List[ValpasOpiskeluoikeus]] = {
+    // TODO: Onko tietokantaluokka oikea paikka tälle?
```

Comment:
Vaikea sanoa, mielestäni ei. Täällä tapahtuu nyt vähän kaikenlaista jopa yllättävää.

Alunperin olin ajatellut että ValpasOppijaService olisi se paikka, jossa koostetaan eri palveluista tuleva data yhteen, ValpasDatabaseService on vain ikkuna tietokantaan ja servlet on vastuussa api-spesifistä asioista (pyynnön parsiminen, http-statukset).

Nyt vastuut ovat vähän sekaisin, kun tietokantapalvelu käyttää muitakin palveluita ja palauttaa suoraan http-statuksia (tämän voi kyllä hyväksyä ihan sillä perusteella, ettei tarvitse mapata erikseen yhdestä sisäisestä virhetyypistä toiseen, jotka lopulta esittävät samaa asiaa). Tämä voi etenkin jatkokehityksen kannalta olla hämmentävää, kun ei ole oikein selkeää paikkaa, mihin mikäkin kuuluu.

Tässä uudessa tavassa on kyllä järkeäkin, koska päästään uudelleenkäyttämään olemassaolevia palikoita. Mutta nyt tämä on tavallaan välitilassa ja mielestäni pitäisi Tehdä Jotain®:

a) Tiputetaan koko OppijaService välistä ja yhdistetään nämä ValpasOppijaRepositoryksi (jumalobjektin vaara)
b) Pidetään database-service kevyenä ja pidetään siellä lähinnä kyselyt. Tuodaan ValidatingAndResolvingExtractor-toiminnallisuus inversion of control -teknologialla sisään oppijaservicen puolelta.

Itse kannattaisin vaihtoehtoa b. Tässä voi olla myös vaarana, että laitetaan tuoleja tuolihuoneeseen ja pöytiä pöytähuoneeseen.